### ISSUE TITLE
Can streamer work with lazy-pages?

### ISSUE THREAD

**Issue by LanYuqiao** [checkpoint-restore/criu-image-streamer#16]
It seems like streaming is incompatible with criu lazy-pages. Any idea on this if I want to achieve lazy restore with streamer ?

**Comment by rst0git** [comment]
> Any idea on this if I want to achieve lazy restore with streamer?

Yes, CRIU currently does not support lazy-pages + image streaming. The implementation would require changes in both CRIU and criu-image-streamer.

**Comment by RohanOnKeys** [comment]
@rst0git can i suggest my thoughts on the implementation for lazy pages ??? ;) or is it already implemented?

**Comment by rst0git** [comment]
> can i suggest my thoughts on the implementation for lazy pages ?

Sure, please feel free to share your thoughts/

>  is it already implemented?

Not yet.

**Comment by RohanOnKeys** [comment]
> > can i suggest my thoughts on the implementation for lazy pages ?
> 
> Sure, please feel free to share your thoughts/
> 
> > is it already implemented?
> 
> Not yet.

how about we leave everything as-is, and add one new optional shard dedicated to lazy page content, only opened when --lazy-pages is passed? keeps the feature "standalone" and the helps avoid messing up rest of the shards

**Comment by RohanOnKeys** [comment]
smth that might look like

On capture: streamer stores incoming lazy pages from CRIU's page-server output into an addressable page cache (in-memory, spill-to-disk if it grows), keyed by page offset.

On serve: streamer speaks just enough of CRIU's page-server wire protocol on that shard to answer out-of-order fetches from the lazy-pages daemon, using whatever has landed in the cache so far, blocking on pages not yet received for the live-migration case.

**Comment by RohanOnKeys** [comment]
Keeps the blast radius small (if things do go wrong)and doesn't touch the capture/serve interface anyone else is relying on.
